### PR TITLE
Prevent Dispose from being called twice on ServiceProvider

### LINKED ISSUE #292: ServiceProvider.Dispose() => Stack overflow
Introduced with bbb9acf. Not actually fixed by 2793fde.

Calls to https://github.com/aspnet/EntityFramework/blob/8667adaecfe8d96cd8929cf83939cae6fd43f189/src/EntityFramework.Core/Internal/DbContextServices.cs#L88
results in stack overflow.

Using Microsoft.Framework.DependencyInjection 1.0.0-beta8-15579

@davidfowl

### PR DESCRIPTION
**Pull request by natemcmaster**
Fix #292, bug introduced in bbb9acf101d4f9b3b818834c60572f67ec16ac63

### CODE REVIEW

File: src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs [hunk 1]
```diff
@@ -211,6 +213,7 @@ private object GetEmptyIEnumerableOrNull(Type serviceType)
 
         private static MethodInfo MonitorEnterMethodInfo = GetMethodInfo<Action<object, bool>>((lockObj, lockTaken) => Monitor.Enter(lockObj, ref lockTaken));
         private static MethodInfo MonitorExitMethodInfo = GetMethodInfo<Action<object>>(lockObj => Monitor.Exit(lockObj));
+        private bool _disposeCalled;
```
Comment (davidfowl) [inline src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs]: move tis to the top of the class

File: test/Microsoft.Framework.DependencyInjection.Tests/ScopingContainerTestBase.cs [hunk 2]
```diff
@@ -122,6 +122,16 @@ public void SelfResolveThenDispose()
         }
 
         [Fact]
+        public void SafelyDisposeNestedProviderReferences()
+        {
+            var container = CreateContainer();
```
Comment (davidfowl) [inline test/Microsoft.Framework.DependencyInjection.Tests/ScopingContainerTestBase.cs]: Where is the assert?

Comment (natemcmaster) [inline test/Microsoft.Framework.DependencyInjection.Tests/ScopingContainerTestBase.cs]: Implicit. This causes stack overflow without the fix.

Comment (davidfowl) [inline test/Microsoft.Framework.DependencyInjection.Tests/ScopingContainerTestBase.cs]: Man more stackoverflows. Crazy :smile:

**Comment by davidfowl** [comment]
:shipit:

**Pull request merged**